src/typechecker/lisp-type.lisp

```diff
@@ -124,3 +130,52 @@ USE-FUNCTION-ENTRIES specifies whether to emit FUNCTION-ENTRY for functions, emi
   (:method ((ty qualified-ty) env)
     (lisp-type (qualified-ty-type ty) env)))
 
+
+(defun try-recurse-transparent-type (from parameter env)
+  "Called when taking a lisp-type of a parameterized type, where a
+concrete type is given to the parameter.  For example, a type `(Foo
+:t)` is defined, and we try to take a lisp-type of `(Foo UFix)`.  The
+`FROM` argument gets the type `Foo` and the `parameter` argument gets
+the type `UFix`.
+
+There is a special opportunity of optimization when `Foo` is defined
+as a transparent type, e.g. `(repr :transparent) (define-type (Foo
+:t) (Content (Lisparray :t)))`.  In that case, `(Foo UFix)` can have a
+Lisp type `(cl:simple-array cl:fixnum)`, which allows Lisp compiler to
+optimize further.  We check the conditions and returns a specialized
+Lisp type if possible.
+
+If we can't get a specialized Lisp type, this returns nil, and let the
+fallback branch take care of it.
+"
+  (labels ((transparent-type? (from)
+             (alexandria:when-let (entry (lookup-type env from :no-error t))
+               (eq (type-entry-explicit-repr entry) :transparent)))
+           (reveal-underlying-type (from)
+             "If `from` type satisfies the optimizable condition, returns two
+values: the 'inner' type (wrappee type) and the 'outer' type (wrapper
+type)."
+             (let ((v (lookup-value-type env from :no-error t)))
+               (when (typep v 'ty-scheme)
+                 (let ((qt (qualified-ty-type (ty-scheme-type v))))
+                   ;; In transparent types, we have this in QT.
+                   ;; (Arrow <innter-type>) -> <outer-type>
+                   (when (function-type-p qt)
+                     (values (function-type-from qt)
+                             (function-type-to qt)))))))
+           )
+    (when (transparent-type? from)
```

Comment:
nit: prefer IF when the return value is meaningful and there are no side effects 

prefer WHEN when the body has side effects or the return value is not
meaningful

Reply:
`(and (transpernt-type? from) ...)` may be clearer.

Reply:
I still think it should be IF, if AND is to be interpreted as a boolean operator and not a secret control flow operator. :)

Reply:
Yeah, in CL, it is doubled as boolean operator and poor man's Maybe monad.  If you're more type oriented, you'd want to distinguish the two.  I'll keep it in mind.